### PR TITLE
Support string repeat with non constant integer

### LINKED ISSUE #2652: Improvement: Support string repeat with a non-constant integer
```python
from lpython import i32

my_string: str = "#"
n: i32 = 5

print(my_string * n)
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
.
.
.
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 3462, in LCompilers::LPython::CommonVisitor<LCompilers::LPython::BodyVisitor>::visit_BinOp(LCompilers::LPython::AST::BinOp_t const&)
    make_BinOp_helper(left, right, op, x.base.base.loc);
  File "/home/saurabh-kumar/Projects/System/lpython/src/lpython/semantics/python_ast_to_asr.cpp", line 2055, in LCompilers::LPython::CommonVisitor<LCompilers::LPython::BodyVisitor>::make_BinOp_helper(LCompilers::ASR::expr_t*, LCompilers::ASR::expr_t*, LCompilers::ASR::binopType, LCompilers::Location const&)
    right_int = ASR::down_cast<ASR::IntegerConstant_t>(
AssertFailed: f != nullptr
```

### PR DESCRIPTION
**Pull request by kmr-srbh**
fixes #2652 

```python
my_string: str = "#"
n: i32 = 5

print(my_string * n)
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
#####
```

### CODE REVIEW

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid the following, unrelated to the changes in the PR, fails:
```python
my_string: str = "#"
repeated_string: str = my_string * -1

print(repeated_string)
```
```console
(base) saurabh-kumar@Awadh:~/Projects/System/lpython$ ./src/bin/lpython ./examples/example.py
malloc(): corrupted top size
Aborted (core dumped)
```

An assert statement had checked for this case before, but interestingly the assertion did not fail. The error is observed when we try to use the obtained value, either by trying to print it, or store it.

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid after you approve these changes, I will make the changes stated in https://github.com/lcompilers/lpython/pull/2651/files#r1573132771

**Comment by kmr-srbh** [comment]
@Shaikh-Ubaid is this assertion here necessarily required?
https://github.com/lcompilers/lpython/blob/3ccdfdb6d6da234a31a0b28fb627691cc9d7e214/src/lpython/semantics/python_ast_to_asr.cpp#L2116

Is the value of `result` prone to bugs? If not, we might as well be able to completely remove the `if (right_is_int) {...}` and `} else if (left_is_int) {...}` blocks as they serve no other purpose then.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -2079,28 +2079,24 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
...
                     ASRUtils::type_get_past_array(right_type));
                 LCOMPILERS_ASSERT(ASRUtils::extract_n_dims_from_ttype(right_type) == 0);
                 left_int = ASR::down_cast<ASR::IntegerConstant_t>(
                                                    ASRUtils::expr_value(left))->m_n;
                 dest_len = right_type2->m_len * left_int;
                 if (dest_len < 0) dest_len = 0;
-                dest_type = ASR::down_cast<ASR::ttype_t>(
-                        ASR::make_Character_t(al, loc, right_type2->m_kind,
-                        dest_len, nullptr));
             }
```
Comment (kmr-srbh) [inline src/lpython/semantics/python_ast_to_asr.cpp:2100]: @Shaikh-Ubaid is this assertion here necessarily required?
https://github.com/lcompilers/lpython/blob/3ccdfdb6d6da234a31a0b28fb627691cc9d7e214/src/lpython/semantics/python_ast_to_asr.cpp#L2116

Is the value of `result` prone to bugs? If not, we might as well be able to completely remove this block as it serves no other purpose than to get `dest_len` for the above stated assertion.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 2]
```diff
@@ -2101,6 +2101,10 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                 dest_type = ASR::down_cast<ASR::ttype_t>(
                         ASR::make_Character_t(al, loc, right_type2->m_kind,
                         dest_len, nullptr));
+            } else if ((left_is_int && ASRUtils::expr_value(left) == nullptr) 
+                || (right_is_int && ASRUtils::expr_value(right) == nullptr)) {
+                dest_type = ASRUtils::TYPE(ASR::make_Character_t(al,
+                            loc, 1, 1, nullptr));
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:2107]:
```suggestion
                dest_type = ASRUtils::TYPE(ASR::make_Character_t(al,
                            loc, 1, -1, nullptr));
```

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 3]
```diff
@@ -2101,6 +2101,10 @@ class CommonVisitor : public AST::BaseVisitor<Struct> {
                 dest_type = ASR::down_cast<ASR::ttype_t>(
                         ASR::make_Character_t(al, loc, right_type2->m_kind,
                         dest_len, nullptr));
+            } else if ((left_is_int && ASRUtils::expr_value(left) == nullptr) 
+                || (right_is_int && ASRUtils::expr_value(right) == nullptr)) {
```
Comment (Shaikh-Ubaid) [inline src/lpython/semantics/python_ast_to_asr.cpp:2105]: Would simply doing the following work?

```suggestion
            } else if (left_is_int || right_is_int) {
```

File: integration_tests/test_str_01.py [hunk 4]
```diff
@@ -58,9 +60,37 @@ def test_str_repeat():
     assert a*3 == "XyzXyzXyz"
     assert a*2*3 == "XyzXyzXyzXyzXyzXyz"
     assert 3*a*3 == "XyzXyzXyzXyzXyzXyzXyzXyzXyz"
-    assert a*-1 == ""
+    # assert a*-1 == ""
```
Comment (Shaikh-Ubaid) [inline integration_tests/test_str_01.py:63]: Can you check if the following fixes this issue?

```diff --git a/src/libasr/runtime/lfortran_intrinsics.c b/src/libasr/runtime/lfortran_intrinsics.c
index ce6c6901a..d6ea89961 100644
--- a/src/libasr/runtime/lfortran_intrinsics.c
+++ b/src/libasr/runtime/lfortran_intrinsics.c
@@ -2140,7 +2140,7 @@ LFORTRAN_API void _lfortran_strrepeat(char** s, int32_t n, char** des
t)
     char* dest_char = (char*)malloc(f_len+trmn_size);
 
     if (s_len == 1) {
-        memset(dest_char, *(*s), n);
+        memset(dest_char, *(*s), f_len);
     } else {
         memcpy(dest_char, *s, s_len);
         int chars_copied = s_len;
```

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 2 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp]: @Shaikh-Ubaid I am curious, why does setting the length to -1 not cause an issue?

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 3 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp:2105]: Thanks for suggesting this simplification @Shaikh-Ubaid ! I realized that as we are checking for compile-time integers above, we might as well have only `else` here. It works.

File: integration_tests/test_str_01.py (hunk 4 above)
Comment (kmr-srbh) [reply integration_tests/test_str_01.py]: Thanks a lot for this @Shaikh-Ubaid ! This works. Interestingly, this was an issue with strings of length 1 only.

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 2 above)
Comment (Shaikh-Ubaid) [reply src/lpython/semantics/python_ast_to_asr.cpp]: Setting `length to 1` could potentially fail in future (for example when there are operations that involve compile-time length). Therefore, it is better to use the appropriate length value in the generated ASR.

`length >= 0` indicates a compile-time length
`length = -1` indicates length is not available and need to be computed at runtime (For example `char*` in c)
`length = -2` indicates length is not available, but the variable is allocatable 
`length = -3` indicates length is available in the form of an expression given as the last parameter of `make_Character_t`.

You can also check the following for more details.

https://github.com/lcompilers/lpython/blob/3ccdfdb6d6da234a31a0b28fb627691cc9d7e214/src/libasr/codegen/asr_to_fortran.cpp#L174-L187

**Review by Shaikh-Ubaid — APPROVED** [review]
Looks good. Thanks for this!

File: src/lpython/semantics/python_ast_to_asr.cpp (hunk 2 above)
Comment (kmr-srbh) [reply src/lpython/semantics/python_ast_to_asr.cpp]: :star_struck: Thanks a lot for this information @Shaikh-Ubaid!

**Pull request merged**